Service/app/models/aci/snapshots.py

```diff
@@ -63,6 +68,104 @@ def before_snapshot_create(data, **kwargs):
             format_timestamp(time.time()))
     return data
 
+def download_snapshot(_id):
+    # download a snapshot .tgz file 
+    snapshot = Snapshots.load(_id=_id)
+    if not snapshot.exists():
+        abort(400, "snapshot %s not found" % _id)
+    else:
+        if snapshot.filename is not None:
+            if os.path.isfile(snapshot.filename) is False:
+                abort(400, "File not found")
+            try: 
+                path = snapshot.filename.split('/')
+                filename = path[len(path) - 1]
+                parent_dir_path = "/".join(path[:-1])
+                return send_from_directory(parent_dir_path,filename, as_attachment = True)
+            except Exception as err:
+                abort(400,str(err))
+    abort(400, "File metadata not found")
+
+def upload_snapshot():
+    from ..utils import get_app_config
+    if request.files is None:
+        abort(400, "No files uploaded")
+    else:
+        tmp_dir = os.path.realpath("%s/%s" % (current_app.config["TMP_DIR"],uuid.uuid4()))
+        config = get_app_config()
+        dst = config.get("DATA_DIR", "/tmp/")
+        try:
+            for filename in request.files:
+                f = request.files[filename] 
+                # validate file is .tgz file and allowed name
+                if not re.search("^[^\/]+\.tgz$", f.filename):
+                    abort(400, "Invalid filename for snapshot: %s" % f.filename)
+                temp_filename = os.path.join(tmp_dir , f.filename)
+                # create the directory
+                os.mkdir(tmp_dir)
+                os.chdir(tmp_dir)
+                # unzip and stuff..
+                f.save(temp_filename)
+                if aci_utils.run_command('tar --force-local -zxf ' + temp_filename) is None:
+                    logger.debug("Traceback:\n %s",traceback.format_exc())
```

Comment:
Rishi - i think we can remove this traceback debug - there won't be a traceback if the tar command fails

Reply:
same on line 112, line 146, and line 150